File: Python/Resize_video.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

def angle_measurement(POSE_PAIRS, p_prof, nPoints):
    teta = list()
    
    for n in range(nPoints):
        partA = np.copy(p_prof[POSE_PAIRS[n][0],:])
        partB = np.copy(p_prof[POSE_PAIRS[n][1],:])
        if((partB[0] - partA[0]) == 0):
            teta.append(90.0)
        else:
            y = partB[1] - partA[1]
            x = partB[0] - partA[0]
            angle = float(np.degrees(np.arctan(y/x)))
            if (angle > 0):
                if (x<0 and y<0):
                    angle += 180
            elif (angle < 0):
                if (x<0 and y>0):
                    angle += 180
            else:
                if (x>0):
                    angle = 0
                elif (x<0):
                    angle = 180
                else:
                    if (y>0):
                        angle = 90
                    if (y<0):
                        angle = -90
            teta.append(angle)
    print(teta)
    return np.array(teta)

def distance_measurement (POSE_PAIRS, p_prof, nPoints):
    d = list()
    for n in range(nPoints):
        partA = np.copy(p_prof[POSE_PAIRS[n][0],:])
        partB = np.copy(p_prof[POSE_PAIRS[n][1],:])
        d.append(float(np.sqrt((partA[0] - partB[0])**2+(partA[1]-partB[1])**2)))
    return np.array(d)

plt.close('all')

POSE_PAIRS = [[1,2], [2,3], [3,4], [1,5], 
             [5,6], [6,7], [1,11], [11,12],
             [12,13], [11,8], [8,9], [9,10],
             [1,0], [0,14], [14,16], [0,15], [15,17]]

nPoints = 17


p_prof = np.genfromtxt("../Photos/victor.txt",delimiter=' ',dtype=None, encoding=None)
p_stud = np.genfromtxt("../Photos/gabriel.txt",delimiter=' ',dtype=None, encoding=None)

frame = cv2.imread("../Photos/leandro.jpg")
frameCopy = np.copy(frame)

d_prof = distance_measurement(POSE_PAIRS, p_prof, nPoints)
a_prof = angle_measurement(POSE_PAIRS, p_prof, nPoints)

d_stud = distance_measurement(POSE_PAIRS, p_stud, nPoints)
a_stud = angle_measurement(POSE_PAIRS, p_stud, nPoints)

dif = 0

for n in range(nPoints):
    partA = np.copy(p_prof[POSE_PAIRS[n][0],:])
    partB = np.copy(p_prof[POSE_PAIRS[n][1],:]) + dif
    
    partB[0] = d_stud[n] * np.cos(np.deg2rad(a_prof[n])) + partA[0]
    partB[1]  = d_stud[n] * np.sin(np.deg2rad(a_prof[n])) + partA[1]

    if(n==0 or n==3 or n==6 or n==9 or n==12 or n==15):
        dif = 0
    else:
        dif = partB - p_prof[POSE_PAIRS[n][1],:]

    p_prof[POSE_PAIRS[n][1],:] = partB
    
dp_prof = distance_measurement(POSE_PAIRS, p_prof, nPoints)
ap_prof = angle_measurement(POSE_PAIRS, p_prof, nPoints)
    
init = np.copy(p_stud[1,:])
fim = np.copy(p_prof[1,:])
offset = fim - init
p_prof -= offset

dataset = pd.DataFrame({'d_stud':d_stud,'dp_prof':dp_prof,
                        'a_stud':a_stud,'ap_prof':ap_prof})

File: C++/resize_video.cpp
#include <opencv2/dnn.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <iostream>
#include <fstream>
#include <unistd.h>
#include <pthread.h>
#include <string>
#include <math.h>

#define PI 3.14159265

using namespace std;
using namespace cv;
using namespace cv::dnn;

const int POSE_PAIRS[17][2] = 
{   
    {1,2}, {2,3}, {3,4},
    {1,5}, {5,6}, {6,7},
    {1,11}, {11,12}, {12,13},
    {11,8}, {8,9}, {9,10},
    {1,0}, {0,14},
    {14,16}, {0,15}, {15,17}
};

int nPoints = 17;

vector<float> angle_measure(vector<Point> points){
    double x, y, angle;
    vector<float> thetas(nPoints);
    for (int n = 0; n < nPoints; n++){
        Point2f partA = points[POSE_PAIRS[n][0]];
        Point2f partB = points[POSE_PAIRS[n][1]];
        if((partB.x - partA.x) == 0){
            thetas[n] = 90;
        }
        else{
            x = partB.x - partA.x;
            y = partB.y - partA.y;
            angle = atan(y/x) * 180 / PI;
            if(angle > 0){
                if(x<0 && y<0){
                    angle += 180;
                }
            }
            else if(angle < 0){
                if(x<0 && y>0){
                    angle +=180;
                }
            }
            else{
                if(x>0){
                    angle = 0;
                }
                else if (x<0){
                    angle = 180;
                }
                else{
                    if (y>0){
                        angle = 90;
                    }
                    if (y<0){
                        angle = -90;
                    }
                }
            }
            thetas[n] = angle;
        }
    }
    return thetas;
}

vector<float> distance_measure(vector<Point> points){
    vector<float> d(nPoints);
    for (int n = 0; n < nPoints; n++){
        Point2f partA = points[POSE_PAIRS[n][0]];
        Point2f partB = points[POSE_PAIRS[n][1]];
        float Ax = (float) partA.x;
        float Ay = (float) partA.y;
        float Bx = (float) partB.x;
        float By = (float) partB.y;
        float x = pow((Ax - Bx),2);
        float y = pow((Ay - By),2);
        d[n] = (float) sqrt(x + y);
    }
    return d;
}

void readfile(vector<Point> *points, string file_path){
    double x, y;

    ifstream ip;
    ip.open(file_path);
    if(!ip.is_open()){
        cout << "file does not exists:" << '\n';
    }
    while(ip.good()){
        ip>>x;
        ip>>y;
        points->push_back(Point((int) x,(int) y));
    }
}

vector<Point> resize_funtion(vector<Point> prof, vector<float> factor, Point2f offset){

    vector<float> a_prof(nPoints);
    vector<float> d_stud(nPoints);
    vector<float> d_prof(nPoints);

    d_prof = distance_measure(prof);
    a_prof = angle_measure(prof);

    for (int n = 0; n < nPoints; n++){
        d_stud[n] = d_prof[n]*factor[n];
    }

    int dif[2] = {0};

    for (int n = 0; n < nPoints; n++){
        Point2f partA = prof[POSE_PAIRS[n][0]];
        Point2f partB = prof[POSE_PAIRS[n][1]];
        partB.x += dif[0];
        partB.y += dif[1];

        partB.x = d_stud[n] * cos(a_prof[n] * PI / 180) + partA.x;
        partB.y = d_stud[n] * sin(a_prof[n] * PI / 180) + partA.y;

        if(n==0 || n==3 || n==6 || n==9 || n==12 || n==15){
            dif[0] = 0;
            dif[1] = 0;
        }
        else{
            Point2f partB_p = prof[POSE_PAIRS[n][1]];
            dif[0] = partB.x - partB_p.x;
            dif[1] = partB.y - partB_p.y;
        }
        prof[POSE_PAIRS[n][1]] = partB;
    }

    vector<float> dp_prof(nPoints);
    vector<float> ap_prof(nPoints);

    dp_prof = distance_measure(prof);
    ap_prof = angle_measure(prof);

    for (int n = 0; n < nPoints+1; n++){
        Point2f p_prof = prof[n];
        p_prof -= offset;
        prof[n] = p_prof;
    }

    return prof;
}

int main(int argc, char **argv){ 

    vector<Point> prof(0);
    vector<Point> stud(0);
    vector<Point> video_prof(0);

    string prof_name = "victor";
    string stud_name = "gabriel";
    string image = "../Photos/" + stud_name + ".jpeg";

    Mat frame = imread(image);

    namedWindow("Output-Skeleton", WINDOW_NORMAL);

    readfile(&stud, "../Photos/" + stud_name + ".txt");
    readfile(&prof, "../Photos/" + prof_name + ".txt");
    readfile(&video_prof, "../Test_files/" + prof_name + ".txt");

    Point2f offset = prof[1] - stud[1];

    vector<float> d_prof(nPoints);
    vector<float> d_stud(nPoints);
    vector<float> factor(nPoints);

    d_prof = distance_measure(prof);
    d_stud = distance_measure(stud);

    for (int n = 0; n < nPoints; n++){
        factor[n] = d_stud[n]/d_prof[n];
        cout << factor[n] << endl;
    }

    int j = 50;

    for (int n = 0; n < nPoints+1; n++){
        prof[n] = video_prof[n + (nPoints+1)*j];
    }

    for (int n = 0; n < nPoints-4; n++){

        Point2f partA = prof[POSE_PAIRS[n][0]];
        Point2f partB = prof[POSE_PAIRS[n][1]];
        Point2f partC = prof[8];
        if (n == 6){
            partC.x = (partB.x + partC.x)/2;
            partC.y = (partB.y + partC.y)/2;
            line(frame, partA, partC, Scalar(0, 255, 255), 1);
        }
        else{
            line(frame, partA, partB, Scalar(0, 255, 255), 1);
            circle(frame, partA, 1, Scalar(0,255,0), -1);
            circle(frame, partB, 1, Scalar(0,255,0), -1);
        }
        circle(frame, partC, 1, Scalar(0,255,0), -1);
    }

    prof = resize_funtion(prof, factor, offset);

    for (int n = 0; n < nPoints-4; n++){

        Point2f partA = prof[POSE_PAIRS[n][0]];
        Point2f partB = prof[POSE_PAIRS[n][1]];
        Point2f partC = prof[8];
        if (n == 6){
            partC.x = (partB.x + partC.x)/2;
            partC.y = (partB.y + partC.y)/2;
            line(frame, partA, partC, Scalar(0, 0, 0), 1);
        }
        else{
            line(frame, partA, partB, Scalar(0, 0, 0), 1);
            circle(frame, partA, 1, Scalar(0,0,255), -1);
            circle(frame, partB, 1, Scalar(0,0,255), -1);
        }
        circle(frame, partC, 1, Scalar(0,0,255), -1);
    }
    cout << prof.size() << endl;

    cout << "imshow" << endl;
    imshow("Output-Skeleton",frame);
    waitKey();

    return 0;
}
